flang/lib/Lower/CharacterExpr.cpp

```diff
@@ -263,7 +265,10 @@ fir::CharBoxValue Fortran::lower::CharacterExprHelper::toScalarCharacter(
 
 mlir::Value Fortran::lower::CharacterExprHelper::createEmbox(
     const fir::CharArrayBoxValue &box) {
-  return createEmbox(toScalarCharacter(box));
+  // Use same embox as for scalar. It's losing the actual data size information
+  // (We do not multiply the length by the array size), but that is what Fortran
+  // call interfaces using boxchar expect.
+  return createEmbox(static_cast<const fir::CharBoxValue &>(box));
```

Comment:
What about the case where an array of CHARACTER is passed to an I/O routine that expects a CHARACTER parameter? There was one of these in fcvs and the output expected the entire array to be interpreted as a string of `n * len` size.

```fortran
CHARACTER(5) :: c(3)
DATA c(1), c(2), c(3) / 'this ', 'is a ', 'string' /
```
and the output was "this is a string" -- 15 characters.

Reply:
Yes, this patch only impact user code interface. I remember that doing this multiplication came from a`materializeCharacterOrSequence` change to accommodate an IO use case. In IO, we are not using  `createEmbox()` but `toScalarCharacter` that is doing this size * lenght multiplication.
https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/lib/Lower/IO.cpp#L558

So we should have no regression here. I will still check that fcvs has no issue and try to add a regression test that exposes the need to do `n*len` in IO to be safe.

Reply:
I confirm this change does not lead to any fcvs regression.
I also found back the fcvs tests that need the IO 'n*len', that  is fm903 that is testing passing array as FORMAT in read/write. 
That is legal as per Fortran 2018 12.6.2.2. point 3. of this sections that tells:
> If default-char-expr is an array, it is treated as if all of the elements of the array were specified in array element
order and were concatenated. 

So `toScalarCharacter()` will need to be updated when we support Fortran pointers (i.e. non-contiguous arrays) since we will actually need to do the concatenation in non-contiguous cases. I have added a TODO  in toScalarCharacter() to keep keep track of that.

Reply:
That all sounds good. Thanks for checking.